### PR TITLE
Add EpisodicStartAverageDiscountedReturnMetric and AverageRewardMetric

### PR DESCRIPTION
**Pull request by le-horizon**
adding two metrics EpisodicStartAverageDiscountedReturnMetric (compute discounted return of the first step of each episode, and then take average) and AverageRewardMetric (compute reward averaged per step in a given episode).

### CODE REVIEW

**Comment by emailweixu** [comment]
> adding some metrics

Please make the title more descriptive and give a brief description of this PR and other PRs. And I feel that many of the PR are not really directly related to lbtq, so "lbtq" should not be part of the title.

**Comment by emailweixu** [comment]
> > adding some metrics
> 
> Please make the title more descriptive and give a brief description of this PR and other PRs. And I feel that many of the PR are not really directly related to lbtq, so "lbtq" should not be part of the title.

In this regard, @breakds did very good job. Perhaps you can take a look how @breakds's PR look like.

File: alf/metrics/metrics.py [hunk 1]
```diff
@@ -408,6 +416,118 @@ def _extract_and_process_acc_value(self, acc, last_episode_indices):
...
+class AverageRewardMetric(AverageDiscountedReturnMetric):
+    """Metric for computing the average reward per time step for each episode.
+    """
+
+    def __init__(self,
+                 example_time_step: TimeStep,
+                 name='AverageReward',
+                 prefix='Metrics',
+                 buffer_size=10):
+        super(AverageRewardMetric, self).__init__(
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: `super().__init__`

File: alf/metrics/metrics.py [hunk 2]
```diff
@@ -408,6 +416,118 @@ def _extract_and_process_acc_value(self, acc, last_episode_indices):
...
+    ground-truth (when rewards are non-negative).
+    """
+
+    def __init__(self,
+                 example_time_step: TimeStep,
+                 name='EpisodicStartAverageDiscountedReturn',
+                 prefix='Metrics',
+                 buffer_size=10,
+                 reward_clip=None):
+        super(EpisodicStartAverageDiscountedReturnMetric, self).__init__(
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: `super().__init__`

File: alf/metrics/metrics.py [hunk 3]
```diff
@@ -408,6 +416,118 @@ def _extract_and_process_acc_value(self, acc, last_episode_indices):
...
+            self._accumulated_discount == 0,
+            torch.ones_like(self._accumulated_discount),
+            self._accumulated_discount)
+        self._accumulated_discount = torch.where(
+            is_first, torch.zeros_like(self._accumulated_discount),
+            self._accumulated_discount)
+
+    def _extract_and_process_acc_value(self, acc, last_episode_indices):
+        """Extract the final accumulated value and divide by the number of
+        steps of an episode.
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: wrong comment ("divide by ...")

File: alf/metrics/metrics.py [hunk 4]
```diff
@@ -408,6 +416,118 @@ def _extract_and_process_acc_value(self, acc, last_episode_indices):
             last_episode_indices]
 
 
+@alf.configurable
+class EpisodicStartAverageDiscountedReturnMetric(
+        AverageDiscountedReturnMetric):
```
Comment (emailweixu) [inline alf/metrics/metrics.py:416]: Why is it a subclass of AverageDiscountedReturnMetric?

File: alf/metrics/metrics.py [hunk 5]
```diff
@@ -408,6 +416,118 @@ def _extract_and_process_acc_value(self, acc, last_episode_indices):
...
+        is added by one otherwise. Therefore, at the episode end, its value
+        equals to episode length - 1.
+
+        Args:
+            time_step (alf.data_structures.TimeStep): batched tensor
+        """
+        is_first = time_step.is_first()
+
+        # update discount for the next time step
+        self._accumulated_discount = (
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: `self._accumulated_discount *= self._discount`

File: alf/metrics/metrics.py [hunk 6]
```diff
@@ -408,6 +416,118 @@ def _extract_and_process_acc_value(self, acc, last_episode_indices):
...
+            self._discount * self._accumulated_discount)
+        self._accumulated_discount = torch.where(
+            self._accumulated_discount == 0,
+            torch.ones_like(self._accumulated_discount),
+            self._accumulated_discount)
+        self._accumulated_discount = torch.where(
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: masked_fill_

File: alf/metrics/metrics.py [hunk 7]
```diff
@@ -325,21 +325,27 @@ def __init__(self,
                  dtype=torch.float32,
                  discount=0.99,
                  reward_transformer=None,
-                 buffer_size=10):
+                 buffer_size=10,
+                 reward_clip=None):
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: `reward_transformer=RewardClipping()` can achieve the same effect.

File: alf/metrics/metrics.py [hunk 8]
```diff
@@ -408,6 +416,118 @@ def _extract_and_process_acc_value(self, acc, last_episode_indices):
...
+
+        ndim = time_step.step_type.ndim
+        if time_step.reward.ndim == ndim:
+            discounted_reward = time_step.reward * self._accumulated_discount
+        else:
+            reward = time_step.reward.reshape(*time_step.step_type.shape, -1)
+            discounted_reward = [
+                reward[..., i] * self._accumulated_discount
+                for i in range(reward.shape[-1])
+            ]
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: `discounted_reward = list(reward * self._accumulated_discount)`

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/metrics/metrics.py (hunk 1 above)
Comment (le-horizon) [reply alf/metrics/metrics.py]: Done

File: alf/metrics/metrics.py (hunk 2 above)
Comment (le-horizon) [reply alf/metrics/metrics.py]: Done

File: alf/metrics/metrics.py (hunk 3 above)
Comment (le-horizon) [reply alf/metrics/metrics.py]: Done

File: alf/metrics/metrics.py (hunk 4 above)
Comment (le-horizon) [reply alf/metrics/metrics.py:416]: This is so that some parameters like discount, reward_clip can be shared.

File: alf/metrics/metrics.py (hunk 5 above)
Comment (le-horizon) [reply alf/metrics/metrics.py]: Done

File: alf/metrics/metrics.py (hunk 6 above)
Comment (le-horizon) [reply alf/metrics/metrics.py]: Done

File: alf/metrics/metrics.py (hunk 7 above)
Comment (le-horizon) [reply alf/metrics/metrics.py]: Removed reward_clip.

File: alf/metrics/metrics.py (hunk 8 above)
Comment (le-horizon) [reply alf/metrics/metrics.py]: Done.  It has to be permuted:
```
            discounted_reward = list(
                (reward * self._accumulated_discount).permute(
                    reward.ndim - 1, *torch.arange(reward.ndim - 1)
                ))
```

**Review by emailweixu — APPROVED** [review]

**Comment by le-horizon** [comment]
closing in favor of https://github.com/HorizonRobotics/alf/pull/1335
